dhall-nixpkgs/Main.hs

```diff
@@ -0,0 +1,828 @@
+toListWith f (Just x ) = f x
+toListWith _  Nothing  = [ ]
+
+nub :: Ord a => [a] -> [a]
+nub = Foldl.fold Foldl.nub
+
+{-| This specialization of `nub` is necessary to work around a type-checking
+    loop
+-}
+nub'
+    :: Ord (f (Fix f)) => [ (Text, Maybe (Fix f)) ] -> [ (Text, Maybe (Fix f)) ]
+nub' = nub
+
+{-| The following Nix code is required reading for understanding how
+    `findExternalDependencies` needs to work:
+
+    <https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/interpreters/dhall/build-dhall-package.nix>
+
+    The Nixpkgs support for Dhall essentially replaces all remote imports with
+    cache hits, but doing so implies that all remote imports must be protected
+    by an integrity check.
+
+    This function finds all remote imports that are transitive dependencies of
+    the given expression, failing if any of them are missing integrity checks.
+-}
+findExternalDependencies :: Expr Src Import -> StateT Status Shell URL
+findExternalDependencies expression = do
+    -- This is a best-effort attempt to pick an import alternative if there is
+    -- more than one
+    let pickAlt (ImportAlt e0 e1)
+            -- If only the latter import has an integrity check, then select
+            -- that
+            | Embed Import{ importHashed = ImportHashed{ hash = Nothing } } <- Dhall.Core.shallowDenote e0
+            , Embed Import{ importHashed = ImportHashed{ hash = Just _  } } <- Dhall.Core.shallowDenote e1 =
+                Just e1
+            -- Otherwise prefer the first import
+            | otherwise =
+                Just e0
+        pickAlt _ =
+            Nothing
+
+    let rewrittenExpression =
+            Dhall.Optics.rewriteOf Dhall.Core.subExpressions pickAlt expression
+
+    import_ <- lift (Turtle.select (Foldable.toList rewrittenExpression))
+
+    parent :| _ <- zoom stack State.get
+
+    child <- hoist liftIO (Dhall.Import.chainImport parent import_)
+
+    let Import{..} = Dhall.Import.chainedImport child
+
+    let ImportHashed{..} = importHashed
+
+    case importMode of
+        Code     -> return ()
+        RawText  -> return ()
+        Location -> empty  -- "as Location" imports aren't real dependencies
+
+    case importType of
+        Missing ->
+            empty
+
+        Env {} ->
+            empty
+
+        Remote url ->
+            case hash of
+                Just _ -> do
+                    return url
+                Nothing -> do
+                    let dependency = Dhall.Core.pretty url
+
+                    die [NeatInterpolation.text|
+Error: Dependency missing a semantic integrity check
+
+Dhall's Nixpkgs requires that all of your remote dependencies are protected by
+by semantic integrity checks.  This ensures that Nix can replace the remote
+imports with cached imports built by Nix instead of the imports being fetched
+via HTTP requests using Dhall.
+
+The following dependency is missing a semantic integrity check:
+
+↳ $dependency
+|]
```

Comment:
IMHO the lack of indentation makes it a bit tricky to understand the structure of the following code. Maybe move these blobs to the top level?